include/deal.II/grid/data_transfer.h

```diff
@@ -0,0 +1,235 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2023 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_tria_data_transfer_h
+#define dealii_tria_data_transfer_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/mpi.h>
+#include <deal.II/base/mpi_large_count.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/grid/cell_status.h>
+#include <deal.II/grid/tria_iterator_selector.h>
+
+#include <boost/range/iterator_range.hpp>
+
+#include <algorithm>
+#include <fstream>
+#include <functional>
+#include <utility>
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+template <int dim, int spacedim = dim>
+DEAL_II_CXX20_REQUIRES((concepts::is_valid_dim_spacedim<dim, spacedim>))
+struct CellAttachedData
+{
+  using cell_iterator = TriaIterator<CellAccessor<dim, spacedim>>;
+
+  /**
+   * Number of functions that get attached to the Triangulation through
+   * register_data_attach() for example SolutionTransfer.
+   */
+  unsigned int n_attached_data_sets;
+
+  /**
+   * Number of functions that need to unpack their data after a call from
+   * load().
+   */
+  unsigned int n_attached_deserialize;
+
+  using pack_callback_t =
+    std::function<std::vector<char>(cell_iterator, CellStatus)>;
+
+  /**
+   * These callback functions will be stored in the order in which they
+   * have been registered with the register_data_attach() function.
+   */
+  std::vector<pack_callback_t> pack_callbacks_fixed;
+  std::vector<pack_callback_t> pack_callbacks_variable;
+};
+
+/**
+ * A structure that stores information about the data that has been, or
+ * will be, attached to cells via the register_data_attach() function
+ * and later retrieved via notify_ready_to_unpack().
+ *
+ * This class in the private scope of parallel::DistributedTriangulationBase
+ * is dedicated to the data transfer across repartitioned meshes
+ * and to/from the file system.
+ *
+ * It is designed to store all data buffers intended for transfer.
+ */
+template <int dim, int spacedim = dim>
+DEAL_II_CXX20_REQUIRES((concepts::is_valid_dim_spacedim<dim, spacedim>))
+class DataTransfer
```

Comment:
Same here: Perhaps move it into an internal namespace?

The name `DataTransfer` is quite general. This was ok while it was inside another class, where context could provide the missing information. But here it would be nice to make the name more specific about what the class does.

Reply:
This is still an issue. Perhaps `CellAttachedDataTransfer`?

Reply:
Renamed to `CellAttachedDataSerializer`, which resembles more the use of the class, and updated the documentation accordingly.